jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -2372,7 +2376,92 @@ else if (oldQueryParams == null || oldQueryParams.size() == 0)
     @Override
     public <T extends HttpUpgradeHandler> T upgrade(Class<T> handlerClass) throws IOException, ServletException
     {
-        throw new ServletException("HttpServletRequest.upgrade() not supported in Jetty");
+        try
+        {
+            HttpConnection connection = (HttpConnection)getAttribute(HttpConnection.class.getName());
+            HttpChannel httpChannel = connection.getHttpChannel();
+            ServletOutputStream outputStream = httpChannel.getResponse().getOutputStream();
+            ServletInputStream inputStream = getInputStream();
+
+            httpChannel.getResponse().setStatus(200); // change the response from state 101 so it can send data back
+
+            AsyncContext asyncContext = forceStartAsync();
+            T t = handlerClass.getDeclaredConstructor().newInstance();
+            t.init(new WebConnection()
+            {
+                @Override
+                public ServletInputStream getInputStream() throws IOException
+                {
+                    return inputStream;
+                }
+
+                @Override
+                public ServletOutputStream getOutputStream() throws IOException
+                {
+                    return outputStream;
+                }
+
+                @Override
+                public void close() throws Exception
+                {
+                    inputStream.close();
+                    outputStream.close();
+                }
+            });
+            connection.addEventListener(new Connection.Listener()
+            {
+                @Override
+                public void onOpened(Connection connection)
+                {
+                }
+
+                @Override
+                public void onClosed(Connection connection)
+                {
+                    try
+                    {
+                        asyncContext.complete();
+                    }
+                    catch (Exception e)
+                    {
+                        LOG.warn("error during upgrade AsyncContext complete", e);
+                    }
+                    try
+                    {
+                        t.destroy();
+                    }
+                    catch (Exception e)
+                    {
+                        LOG.warn("error during upgrade HttpUpgradeHandler destroy", e);
+                    }
+                    connection.removeEventListener(this);
+                }
+            });
+
+            connection.getParser().servletUpgrade(); // tell the parser it's now parsing content
```

Comment:
I think we might also need to twiddle the generator here to make sure it doesn't do chunking

Reply:
Done, but required more intrusive hacking.